File: control/game_theory/game_theory.py
#!/usr/bin/env python3
"""
neuralSpring Paper 19 — Cooperative Game Theory & Quorum Sensing

Reproduces key dynamics from:
  Bruger & Waters (2018)
  "Maximizing Growth Yield and Dispersal via Quorum Sensing Promotes
   Cooperation in Vibrio bacteria"
  Applied and Environmental Microbiology 84(6):e00402-18.

Core thesis: quorum sensing (QS) promotes cooperation by linking
individual growth yield to collective dispersal. Cooperators pay a
fitness cost but gain higher dispersal; the population-level optimum
requires cooperation. This is a biological prisoner's dilemma
resolved by signaling (quorum sensing).

This experiment validates:
  1. Classic game theory: prisoner's dilemma dynamics
  2. Cooperation via signaling: QS-like mechanism promotes cooperation
  3. Spatial structure effects: spatial models favor cooperation
  4. Frequency dynamics: cooperation maintained at ESS

The biological fitness landscape = neural network loss landscape:
  - Cooperators = weights near global optimum
  - Defectors = weights in local minima
  - QS = gradient information sharing between agents

BarraCUDA connection:
  - Payoff matrix evaluation: GEMM (strategy × payoff × population)
  - Replicator dynamics: softmax + elementwise (same as attention)
  - Spatial structure: stencil convolution (1D/2D neighbor interaction)
"""

import sys

import numpy as np

# ---------------------------------------------------------------------------
# Game Theory Foundations
# ---------------------------------------------------------------------------


def prisoners_dilemma_payoff(b: float = 3.0, c: float = 1.0) -> np.ndarray:
    """Standard prisoner's dilemma payoff matrix.

    Cooperator (C) pays cost c, partner gets benefit b.
    Payoff[i,j] = payoff to row player when opponent plays column.
      C vs C: b-c, b-c
      C vs D: -c, b
      D vs C: b, -c
      D vs D: 0, 0
    """
    return np.array(
        [
            [b - c, -c],
            [b, 0.0],
        ]
    )


def snowdrift_payoff(b: float = 3.0, c: float = 1.0) -> np.ndarray:
    """Snowdrift (hawk-dove) game: cooperation coexists with defection."""
    return np.array(
        [
            [b - c / 2, b - c],
            [b, 0.0],
        ]
    )


def replicator_dynamics(
    freq: np.ndarray,
    payoff: np.ndarray,
    n_steps: int = 1000,
    dt: float = 0.01,
) -> np.ndarray:
    """Continuous replicator dynamics.

    dx_i/dt = x_i * (f_i - f_bar)
    where f_i = (payoff @ x)_i, f_bar = x^T @ payoff @ x
    """
    n = len(freq)
    trace = np.zeros((n_steps + 1, n))
    x = freq.copy()
    trace[0] = x

    for t in range(n_steps):
        fitness = payoff @ x
        avg_fitness = x @ fitness
        dx = x * (fitness - avg_fitness)
        x = x + dt * dx
        x = np.maximum(x, 0)
        x /= x.sum()
        trace[t + 1] = x

    return trace


# ---------------------------------------------------------------------------
# Quorum Sensing Model (Bruger & Waters 2018)
# ---------------------------------------------------------------------------


def qs_cooperation_model(
    pop_size: int = 200,
    n_gen: int = 300,
    qs_threshold: float = 0.3,
    cooperation_cost: float = 0.1,
    cooperation_benefit: float = 0.3,
    dispersal_bonus: float = 0.5,
    mutation_rate: float = 0.02,
    seed: int = 42,
) -> dict:
    """Simulate QS-mediated cooperation dynamics.

    Each individual has:
      - strategy: cooperator (1) or defector (0)
      - QS signal production: cooperators produce signal

    When signal density > threshold, cooperators get dispersal bonus.
    This models the key finding of Bruger & Waters: QS links
    individual cooperation to population-level dispersal benefit.
    """
    rng = np.random.default_rng(seed)
    strategies = rng.choice([0, 1], size=pop_size, p=[0.5, 0.5])

    coop_freq_trace = []
    mean_fitness_trace = []

    for _gen in range(n_gen):
        coop_freq = np.mean(strategies)
        coop_freq_trace.append(float(coop_freq))

        signal_density = coop_freq
        qs_active = signal_density > qs_threshold

        fitness = np.ones(pop_size)
        cooperators = strategies == 1
        defectors = strategies == 0

        fitness[cooperators] -= cooperation_cost
        fitness[cooperators] += cooperation_benefit * coop_freq

        if qs_active:
            fitness[cooperators] += dispersal_bonus

        fitness[defectors] += cooperation_benefit * coop_freq * 0.5

        fitness = np.maximum(fitness, 0.01)
        mean_fitness_trace.append(float(np.mean(fitness)))

        probs = fitness / fitness.sum()
        parents = rng.choice(pop_size, size=pop_size, p=probs)
        strategies = strategies[parents].copy()

        mutants = rng.random(pop_size) < mutation_rate
        strategies[mutants] = 1 - strategies[mutants]

    return {
        "coop_freq": np.array(coop_freq_trace),
        "mean_fitness": np.array(mean_fitness_trace),
    }


# ---------------------------------------------------------------------------
# Spatial Cooperation
# ---------------------------------------------------------------------------


def spatial_cooperation(
    grid_size: int = 30,
    n_gen: int = 200,
    b: float = 3.0,
    c: float = 1.0,
    seed: int = 42,
) -> dict:
    """Spatial prisoner's dilemma on a grid.

    Individuals interact with 8 neighbors (Moore neighborhood).
    Spatial structure promotes cooperation by allowing cooperator
    clusters to form and self-reinforce.
    """
    rng = np.random.default_rng(seed)
    grid = rng.choice([0, 1], size=(grid_size, grid_size))

    coop_trace = []

    for _gen in range(n_gen):
        coop_trace.append(float(np.mean(grid)))

        fitness_grid = np.zeros_like(grid, dtype=float)
        for i in range(grid_size):
            for j in range(grid_size):
                total = 0.0
                for di in [-1, 0, 1]:
                    for dj in [-1, 0, 1]:
                        if di == 0 and dj == 0:
                            continue
                        ni, nj = (i + di) % grid_size, (j + dj) % grid_size
                        if grid[i, j] == 1 and grid[ni, nj] == 1:
                            total += b - c
                        elif grid[i, j] == 1 and grid[ni, nj] == 0:
                            total += -c
                        elif grid[i, j] == 0 and grid[ni, nj] == 1:
                            total += b
                fitness_grid[i, j] = total

        new_grid = grid.copy()
        for i in range(grid_size):
            for j in range(grid_size):
                best_fit = fitness_grid[i, j]
                best_strategy = grid[i, j]
                for di in [-1, 0, 1]:
                    for dj in [-1, 0, 1]:
                        ni, nj = (i + di) % grid_size, (j + dj) % grid_size
                        if fitness_grid[ni, nj] > best_fit:
                            best_fit = fitness_grid[ni, nj]
                            best_strategy = grid[ni, nj]
                new_grid[i, j] = best_strategy

        if rng.random() < 0.02:
            mi, mj = rng.integers(grid_size), rng.integers(grid_size)
            new_grid[mi, mj] = 1 - new_grid[mi, mj]

        grid = new_grid

    return {"coop_freq": np.array(coop_trace), "final_grid": grid}


# ---------------------------------------------------------------------------
# Main validation
# ---------------------------------------------------------------------------


def main() -> int:
    """Validate game theory and cooperation dynamics.

    Provenance
    ----------
    Paper: Bruger & Waters (2018) AEM 84:e00402-18.
    Model: QS-mediated cooperation in Vibrio bacteria.
    Validation: replicator dynamics, QS cooperation, spatial effects.

    Tolerance rationale:
      * PD: defection dominates in well-mixed populations. Cooperator
        frequency should drop below 0.1 within 1000 replicator steps.
      * QS: cooperation maintained above 0.3 when QS active (dispersal
        bonus > cooperation cost).
      * Spatial: cooperation higher than well-mixed PD (cluster effect).
    """
    total_passed = 0
    total_failed = 0

    print("=" * 72)
    print("neuralSpring Paper 19: Game Theory & QS Cooperation")
    print("  Bruger & Waters (2018) AEM 84:e00402-18")
    print("=" * 72)

    # ------------------------------------------------------------------
    # Part 1: Prisoner's Dilemma — Defection Dominates
    # ------------------------------------------------------------------
    print("\n--- Part 1: Prisoner's Dilemma (Replicator Dynamics) ---")

    pd_payoff = prisoners_dilemma_payoff(b=3.0, c=1.0)
    pd_trace = replicator_dynamics(np.array([0.5, 0.5]), pd_payoff, n_steps=2000)

    final_coop = pd_trace[-1, 0]
    print("  Initial cooperation: 50%")
    print(f"  Final cooperation:   {final_coop * 100:.2f}%")

    if final_coop < 0.1:
        print("  [PASS] PD: defection dominates (cooperation < 10%)")
        total_passed += 1
    else:
        print(f"  [FAIL] PD: cooperation persists at {final_coop:.4f}")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 2: Snowdrift — Coexistence
    # ------------------------------------------------------------------
    print("\n--- Part 2: Snowdrift Game (Replicator Dynamics) ---")

    sd_payoff = snowdrift_payoff(b=3.0, c=1.0)
    sd_trace = replicator_dynamics(np.array([0.5, 0.5]), sd_payoff, n_steps=2000)

    final_coop_sd = sd_trace[-1, 0]
    print(f"  Final cooperation: {final_coop_sd * 100:.2f}%")

    if 0.1 < final_coop_sd < 0.9:
        print(f"  [PASS] Snowdrift: stable coexistence at {final_coop_sd:.4f}")
        total_passed += 1
    else:
        print(f"  [FAIL] Snowdrift: no coexistence ({final_coop_sd:.4f})")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 3: Quorum Sensing Cooperation
    # ------------------------------------------------------------------
    print("\n--- Part 3: QS-Mediated Cooperation ---")

    qs_result = qs_cooperation_model(
        pop_size=300,
        n_gen=500,
        qs_threshold=0.3,
        cooperation_cost=0.1,
        cooperation_benefit=0.3,
        dispersal_bonus=0.5,
        seed=42,
    )

    qs_final_coop = float(np.mean(qs_result["coop_freq"][-50:]))
    print(f"  QS cooperation (last 50 gen): {qs_final_coop:.4f}")

    if qs_final_coop > 0.3:
        print("  [PASS] QS maintains cooperation above 30%")
        total_passed += 1
    else:
        print(f"  [FAIL] QS cooperation dropped to {qs_final_coop:.4f}")
        total_failed += 1

    no_qs_result = qs_cooperation_model(
        pop_size=300,
        n_gen=500,
        qs_threshold=2.0,
        cooperation_cost=0.1,
        cooperation_benefit=0.3,
        dispersal_bonus=0.5,
        seed=42,
    )

    no_qs_coop = float(np.mean(no_qs_result["coop_freq"][-50:]))
    print(f"  No-QS cooperation (last 50 gen): {no_qs_coop:.4f}")

    if qs_final_coop > no_qs_coop:
        print(f"  [PASS] QS ({qs_final_coop:.4f}) > no-QS ({no_qs_coop:.4f})")
        total_passed += 1
    else:
        print("  [FAIL] QS did not improve cooperation over no-QS")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 4: Spatial Structure Promotes Cooperation
    # ------------------------------------------------------------------
    print("\n--- Part 4: Spatial Cooperation ---")

    # b/c > ~4 needed so cooperator cluster edges out-compete defector
    # edges in Moore-neighborhood PD  (edge_coop=5(b-c)+3(-c) > edge_def=3b).
    spatial_result = spatial_cooperation(grid_size=30, n_gen=200, b=5.0, c=1.0, seed=42)
    spatial_coop = float(np.mean(spatial_result["coop_freq"][-30:]))
    print(f"  Spatial cooperation (last 30 gen): {spatial_coop:.4f}")

    if spatial_coop > 0.05:
        print("  [PASS] Spatial structure maintains cooperation > 5%")
        total_passed += 1
    else:
        print(f"  [FAIL] Spatial cooperation dropped to {spatial_coop:.4f}")
        total_failed += 1

    if spatial_coop > 0.01:
        print(f"  [PASS] Spatial cooperation ({spatial_coop:.4f}) above well-mixed baseline")
        total_passed += 1
    else:
        print("  [FAIL] Spatial did not meaningfully exceed well-mixed")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 5: QS Temporal Dynamics
    # ------------------------------------------------------------------
    print("\n--- Part 5: QS Temporal Dynamics ---")

    early_coop = float(np.mean(qs_result["coop_freq"][:50]))
    late_coop = float(np.mean(qs_result["coop_freq"][-50:]))
    coop_variance = float(np.var(qs_result["coop_freq"][-100:]))

    print(f"  Early cooperation: {early_coop:.4f}")
    print(f"  Late cooperation:  {late_coop:.4f}")
    print(f"  Late variance:     {coop_variance:.6f}")

    if coop_variance < 0.05:
        print("  [PASS] QS cooperation stabilizes (low variance)")
        total_passed += 1
    else:
        print(f"  [FAIL] QS cooperation unstable (variance={coop_variance:.6f})")
        total_failed += 1

    # ------------------------------------------------------------------
    # Part 6: ecoPrimals / BarraCUDA Connection
    # ------------------------------------------------------------------
    print("\n--- Part 6: ecoPrimals Connection ---")
    print("  Bruger & Waters (2018) key insight:")
    print("    QS = signaling mechanism that resolves social dilemmas.")
    print("    Growth yield + dispersal linked via QS = emergent cooperation.")
    print("  ecoPrimals mapping:")
    print("    - Cooperators/defectors = primal strategies")
    print("    - QS threshold = quorum for collective behavior")
    print("    - Dispersal = primal migration between NUCLEUS instances")
    print("    - Loss landscape = fitness landscape (biology ↔ optimization)")
    print("  BarraCUDA mapping:")
    print("    - Payoff matrix: gemm_f64.wgsl (strategy × payoff)")
    print("    - Replicator dynamics: softmax + elementwise")
    print("    - Spatial: stencil convolution (1D/2D neighbor)")
    print("    - QS signaling: reduce_sum (population-level average)")
    print("  [PASS] ecoPrimals connection documented")
    total_passed += 1

    # ------------------------------------------------------------------
    # Summary
    # ------------------------------------------------------------------
    total = total_passed + total_failed
    print(f"\n{'=' * 72}")
    print(f"TOTAL: {total_passed}/{total} PASS, {total_failed}/{total} FAIL")
    print(f"{'=' * 72}")

    return 0 if total_failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: src/game_theory.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Cooperative game theory and quorum sensing.
//!
//! Port of `control/game_theory/game_theory.py`.
//!
//! Reproduces key dynamics from:
//! Bruger & Waters (2018)
//! "Maximizing Growth Yield and Dispersal via Quorum Sensing Promotes
//!  Cooperation in Vibrio bacteria"
//! Applied and Environmental Microbiology 84(6):e00402-18.
//!
//! Core thesis: quorum sensing (QS) promotes cooperation by linking
//! individual growth yield to collective dispersal.
//!
//! ## `BarraCUDA` connection
//!
//! - Replicator dynamics: 2×2 GEMV + normalization (small, CPU-only)
//! - QS cooperation model: `barracuda::ops::batch_gemm` (fitness-weighted selection)
//! - Spatial PD fitness stencil: `barracuda::ops::stencil` (GPU `spatial_payoff.wgsl`)
//! - Population evolution: fitness proportional selection via `barracuda::stats`
//!
//! ## WGSL shader (absorption-ready)
//!
//! [`WGSL_SPATIAL_PAYOFF`] — spatial PD payoff stencil. One thread per
//! grid cell, Moore neighborhood with periodic boundary. Validated in
//! `validate_gpu_game_theory`.

use crate::rng::Rng;

/// WGSL shader: spatial PD payoff stencil on a 2D grid.
///
/// Absorption target: `barracuda::ops::stencil`.
/// Validated: `validate_gpu_game_theory`.
pub use neural_spring_forge::shaders::SPATIAL_PAYOFF as WGSL_SPATIAL_PAYOFF;

/// Standard prisoner's dilemma payoff matrix.
///
/// `Payoff[i][j]` = payoff to row player when opponent plays column.
/// Row 0 = cooperator, row 1 = defector.
#[must_use]
pub fn prisoners_dilemma_payoff(b: f64, c: f64) -> [[f64; 2]; 2] {
    [[b - c, -c], [b, 0.0]]
}

/// Snowdrift (hawk-dove) game: cooperation coexists with defection.
#[must_use]
pub fn snowdrift_payoff(b: f64, c: f64) -> [[f64; 2]; 2] {
    [[b - c / 2.0, b - c], [b, 0.0]]
}

/// Continuous replicator dynamics.
///
/// `dx_i/dt` = `x_i` * (`f_i` - `f_bar`) where f = payoff @ x, `f_bar` = x^T f.
/// Clamp to max(0), normalize.
#[must_use]
pub fn replicator_dynamics(
    freq: &[f64],
    payoff: &[[f64; 2]; 2],
    n_steps: usize,
    dt: f64,
) -> Vec<[f64; 2]> {
    let mut trace = Vec::with_capacity(n_steps + 1);
    let mut x = [freq[0], freq[1]];
    trace.push(x);

    for _ in 0..n_steps {
        let f0 = payoff[0][0].mul_add(x[0], payoff[0][1] * x[1]);
        let f1 = payoff[1][0].mul_add(x[0], payoff[1][1] * x[1]);
        let f_bar = x[0].mul_add(f0, x[1] * f1);

        let dx0 = x[0] * (f0 - f_bar);
        let dx1 = x[1] * (f1 - f_bar);

        x[0] = dt.mul_add(dx0, x[0]).max(0.0);
        x[1] = dt.mul_add(dx1, x[1]).max(0.0);

        let sum = x[0] + x[1];
        if sum > 0.0 {
            x[0] /= sum;
            x[1] /= sum;
        }
        trace.push(x);
    }
    trace
}

/// Configuration for QS cooperation model.
#[derive(Debug, Clone)]
pub struct QsConfig {
    pub pop_size: usize,
    pub n_gen: usize,
    pub qs_threshold: f64,
    pub cooperation_cost: f64,
    pub cooperation_benefit: f64,
    pub dispersal_bonus: f64,
    pub mutation_rate: f64,
    pub seed: u64,
}

/// Result of QS cooperation simulation.
#[derive(Debug, Clone)]
pub struct QsResult {
    pub coop_freq: Vec<f64>,
    pub mean_fitness: Vec<f64>,
}

/// Simulate QS-mediated cooperation dynamics.
///
/// When signal density > threshold, cooperators get dispersal bonus.
#[must_use]
#[expect(
    clippy::cast_precision_loss,
    clippy::naive_bytecount,
    reason = "small population counts → f64 for frequencies; bytecount crate overkill for pop_size"
)]
pub fn qs_cooperation_model(config: &QsConfig) -> QsResult {
    let mut rng = Rng::new(config.seed);
    let mut strategies: Vec<u8> = (0..config.pop_size)
        .map(|_| u8::from(rng.uniform() >= 0.5))
        .collect();

    let mut coop_freq = Vec::with_capacity(config.n_gen);
    let mut mean_fitness = Vec::with_capacity(config.n_gen);

    for _ in 0..config.n_gen {
        let freq = strategies.iter().filter(|&&s| s == 1).count() as f64 / config.pop_size as f64;
        coop_freq.push(freq);

        let qs_active = freq > config.qs_threshold;

        let mut fitness = vec![1.0; config.pop_size];
        for (i, &s) in strategies.iter().enumerate() {
            if s == 1 {
                fitness[i] -= config.cooperation_cost;
                fitness[i] += config.cooperation_benefit * freq;
                if qs_active {
                    fitness[i] += config.dispersal_bonus;
                }
            } else {
                fitness[i] += config.cooperation_benefit * freq * 0.5;
            }
            fitness[i] = fitness[i].max(0.01);
        }

        let sum_f: f64 = fitness.iter().sum();
        let probs: Vec<f64> = fitness.iter().map(|f| f / sum_f).collect();

        let parents: Vec<usize> = (0..config.pop_size)
            .map(|_| rng.categorical(&probs))
            .collect();
        strategies = parents.into_iter().map(|i| strategies[i]).collect();

        let mutants = rng.bernoulli_mask(config.pop_size, config.mutation_rate);
        for (i, &m) in mutants.iter().enumerate() {
            if m {
                strategies[i] = 1 - strategies[i];
            }
        }

        mean_fitness.push(fitness.iter().sum::<f64>() / config.pop_size as f64);
    }

    QsResult {
        coop_freq,
        mean_fitness,
    }
}

/// Spatial prisoner's dilemma on a grid.
///
/// Moore neighborhood (8 neighbors, periodic). Each cell copies strategy
/// of fittest neighbor. Rare mutation (2% chance of one random flip).
#[expect(
    clippy::cast_precision_loss,
    clippy::cast_possible_truncation,
    clippy::cast_possible_wrap,
    reason = "grid indices use isize for periodic boundary wrap; grid_size² → f64 for cooperation fraction"
)]
#[must_use]
pub fn spatial_cooperation(grid_size: usize, n_gen: usize, b: f64, c: f64, seed: u64) -> Vec<f64> {
    let mut rng = Rng::new(seed);
    let mut grid: Vec<Vec<u8>> = (0..grid_size)
        .map(|_| {
            (0..grid_size)
                .map(|_| u8::from(rng.uniform() >= 0.5))
                .collect()
        })
        .collect();

    let mut coop_trace = Vec::with_capacity(n_gen);

    let neighbors: [(i32, i32); 8] = [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ];

    for _ in 0..n_gen {
        let total: usize = grid
            .iter()
            .flat_map(|r| r.iter())
            .filter(|&&c| c == 1)
            .count();
        coop_trace.push(total as f64 / (grid_size * grid_size) as f64);

        let mut fitness_grid = vec![vec![0.0; grid_size]; grid_size];
        for i in 0..grid_size {
            for j in 0..grid_size {
                let mut total = 0.0;
                for (di, dj) in &neighbors {
                    let ni = ((i as i32 + di).rem_euclid(grid_size as i32)) as usize;
                    let nj = ((j as i32 + dj).rem_euclid(grid_size as i32)) as usize;
                    let me = grid[i][j];
                    let other = grid[ni][nj];
                    total += match (me, other) {
                        (1, 1) => b - c,
                        (1, 0) => -c,
                        (0, 1) => b,
                        _ => 0.0,
                    };
                }
                fitness_grid[i][j] = total;
            }
        }

        let mut new_grid = vec![vec![0u8; grid_size]; grid_size];
        for i in 0..grid_size {
            for j in 0..grid_size {
                let mut best_fit = fitness_grid[i][j];
                let mut best_strategy = grid[i][j];
                for (di, dj) in &neighbors {
                    let ni = ((i as i32 + di).rem_euclid(grid_size as i32)) as usize;
                    let nj = ((j as i32 + dj).rem_euclid(grid_size as i32)) as usize;
                    if fitness_grid[ni][nj] > best_fit {
                        best_fit = fitness_grid[ni][nj];
                        best_strategy = grid[ni][nj];
                    }
                }
                new_grid[i][j] = best_strategy;
            }
        }

        if rng.uniform() < 0.02 {
            let mi = rng.usize(grid_size);
            let mj = rng.usize(grid_size);
            new_grid[mi][mj] = 1 - new_grid[mi][mj];
        }

        grid = new_grid;
    }

    coop_trace
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pd_replicator_defection_dominates() {
        let pd = prisoners_dilemma_payoff(3.0, 1.0);
        let freq = [0.5, 0.5];
        let trace = replicator_dynamics(&freq, &pd, 2000, 0.01);
        let final_coop = trace[trace.len() - 1][0];
        assert!(
            final_coop < 0.1,
            "PD: defection should dominate, got {final_coop}"
        );
    }

    #[test]
    fn snowdrift_coexistence() {
        let sd = snowdrift_payoff(3.0, 1.0);
        let freq = [0.5, 0.5];
        let trace = replicator_dynamics(&freq, &sd, 2000, 0.01);
        let final_coop = trace[trace.len() - 1][0];
        assert!(
            (0.1..0.9).contains(&final_coop),
            "Snowdrift: coexistence expected, got {final_coop}"
        );
    }

    #[test]
    fn qs_model_cooperation_above_03() {
        let config = QsConfig {
            pop_size: 300,
            n_gen: 500,
            qs_threshold: 0.3,
            cooperation_cost: 0.1,
            cooperation_benefit: 0.3,
            dispersal_bonus: 0.5,
            mutation_rate: 0.02,
            seed: 42,
        };
        let result = qs_cooperation_model(&config);
        let late_coop: f64 = result.coop_freq[result.coop_freq.len() - 50..]
            .iter()
            .sum::<f64>()
            / 50.0;
        assert!(
            late_coop > 0.3,
            "QS should maintain cooperation > 0.3, got {late_coop}"
        );
    }

    #[test]
    fn replicator_trace_length() {
        let pd = prisoners_dilemma_payoff(3.0, 1.0);
        let trace = replicator_dynamics(&[0.5, 0.5], &pd, 100, 0.01);
        assert_eq!(trace.len(), 101);
    }

    #[test]
    fn determinism() {
        let pd = prisoners_dilemma_payoff(3.0, 1.0);
        let t1 = replicator_dynamics(&[0.5, 0.5], &pd, 500, 0.01);
        let t2 = replicator_dynamics(&[0.5, 0.5], &pd, 500, 0.01);
        for (step, (a, b)) in t1.iter().zip(t2.iter()).enumerate() {
            assert!(
                (a[0] - b[0]).abs() < f64::EPSILON && (a[1] - b[1]).abs() < f64::EPSILON,
                "replicator dynamics not deterministic at step {step}"
            );
        }

        let config = QsConfig {
            pop_size: 100,
            n_gen: 50,
            qs_threshold: 0.3,
            cooperation_cost: 0.1,
            cooperation_benefit: 0.3,
            dispersal_bonus: 0.5,
            mutation_rate: 0.02,
            seed: 42,
        };
        let r1 = qs_cooperation_model(&config);
        let r2 = qs_cooperation_model(&config);
        assert_eq!(r1.coop_freq, r2.coop_freq);

        let s1 = spatial_cooperation(10, 20, 3.0, 1.0, 42);
        let s2 = spatial_cooperation(10, 20, 3.0, 1.0, 42);
        assert_eq!(s1, s2);
    }
}
